src/tribler-common/tribler_common/utilities.py

```diff
@@ -27,6 +29,46 @@ def uri_to_path(uri):
 
 
 fts_query_re = re.compile(r'\w+', re.UNICODE)
+extract_tags_re = re.compile(r'\[\w+\]')
+
+
+@dataclass
+class Query:
+    original_query: Optional[str]
+    tags: Set[str] = field(default_factory=set)
+    fts_text: str = ''
+
+
+def parse_query(query: Optional[str]) -> Query:
+    """
+    The query structure:
+        query = [tag1][tag2] text
+                 ^           ^
+                tags        fts query
+    """
```

Comment:
> The function extracts tags from the query string. But why should tags be inside the query string at all?

Because this way users can specify "tags for search" by using the current UI. See: https://github.com/Tribler/tribler/issues/6551

We haven't a separate UI for specifying tags for search.

> If we really need to put tags into the query string, we probably need to use another syntax to separate tags, as the current algorithm that extracts tags from a query string can remove legitimate parts of the title and affect the full-text search result.

Square brackets are a natural way to specify tags inside a query string. 
![image](https://user-images.githubusercontent.com/13798583/145375305-b99911f4-0288-447e-b579-9ea44a4b8651.png)

> It is not uncommon to encounter a torrent title that starts with a square bracket, in the format [section_name] rest of the title. If a user copies and pastes such a torrent title into a query field, the current algorithm will treat the section_name as a tag name and produce incorrect results.

This is a valid point. Probably to specify "tags for search" we can use double square brackets.

`[[tag1]] [[tag2]] some query with [brackets]`

> After some experiments, I think I found a syntax that does not interfere with torrent titles: \tag-1\tag-2\query string

I appreciate your effort and spent time, but for me, this syntax looks unnatural. I never see in any other applications tags to be specified this way.

Reply:
> If a user copies and pastes such a torrent title into a query field, the current algorithm will treat the section_name as a tag name and produce incorrect results.

No, you are a bit wrong in this. In the case that you describe only FTS will be performed.

See: https://github.com/drew2a/tribler/blob/465c7be7b093dd9708d96c3ea1b84198208083a7/src/tribler-gui/tribler_gui/tribler_window.py#L1026-L1054

Reply:
But you are right, the search will not be accurate, as in this case tags at the beginning of the query will be removed. I'll fix it.

Reply:
For further discussion: in case of the query `'some query with [brackets]'`, the real query will be: `{'txt_filter': '"some" "query" "with" "brackets"*', 'hide_xxx': True}`

See:  https://github.com/Tribler/tribler/blob/3419e31553b0525717b9b9bc55a9cb8f6b81b659/src/tribler-gui/tribler_gui/widgets/searchresultswidget.py#L118

Reply:
Solved by replacing square brackets with the `#` character.